robolectric/src/test/java/org/robolectric/shadows/ShadowBitmapTest.java

```diff
@@ -822,4 +822,19 @@ private void createScaledBitmap_drawOnScaled(boolean filter) {
     p.setColor(Color.BLACK);
     canvas.drawRect(new Rect(0, 0, 32, 32), p);
   }
+
+  @Test
+  public void drawRect_toCanvas() {
```

Comment:
1. I think this test should moved to `ShadowCanvasTest.java`.
2. This test only ensure `Canva#drawRect(Rect rect, Paint paint)` works on Robolectric, but it doesn't ensure `Canvas#drawRect(float left, float top, float right, float bottom, Paint paint)` works as expected, and it is shown by origin issue.
3. From https://github.com/robolectric/robolectric/blob/183437a0b45ff97f2df3128ec01896ffaef5e737/shadows/framework/src/main/java/org/robolectric/shadows/ShadowCanvas.java#L243-L256, the shadow of `Canva#drawRect(float left, float top, float right, float bottom, Paint paint)` doesn't update target `Bitmap` content as `Canvas#drawRect(Rect rect, Paint paint)` did. IMO, you could improve this shadow implementation and add related test for it at `ShadowCanvasTest.java`.

Reply:
Yeah, the goal here was to fix Robolectric in a way that the original test passes.

There were some nuances to this issue that I did not think about initially.

There are two variants of Canvas.drawRect-- one if them is `drawRect(Rect r, Paint paint)`, and the other is `drawRect(float left, float top, float right, float bottom, Paint paint)`. Currently, in Robolectric,  [drawRect(Rect r, Paint paint)](https://github.com/robolectric/robolectric/blob/master/shadows/framework/src/main/java/org/robolectric/shadows/ShadowCanvas.java#L252-L255) has a more correct shadow implementation, in that it calls into ShadowBitmap.drawRect to do actual drawing.

The `drawRect(float left, float top, float right, float bottom, Paint paint)` does not call into ShadowBitmap, so it is effectively a no-op. The goal is to update it to call into ShadowBitmap to do some drawing.

One nuance is that `Rect` in Android is a rect of ints, and `drawRect(float left, float top, float right, float bottom, Paint paint)` takes floats. So I don't think calling the existing ShadowBitmap.drawRect will work here (unless all of the floats have 0 in the fractional part).

One possibility is to implement a new ShadowBitmap.drawRect that takes a RectF and uses AWT:

```
  void drawRect(RectF r, Paint paint) {
    if (bufferedImage == null) {
      return;
    }
    Graphics2D graphics2D = bufferedImage.createGraphics();
    Rectangle2D r2d = new Rectangle2D.Float(r.left, r.top, r.right - r.left, r.bottom - r.top);
    graphics2D.draw(r2d);
    graphics2D.dispose();
  }
``` 

Then you could call that from ShadowCanvas using:

```
    if (targetBitmap != null) {
      ShadowBitmap shadowTargetBitmap = Shadows.shadowOf(targetBitmap);
      shadowTargetBitmap.drawRect(new RectF(left, top, right, bottom), paint);
    }
```

I vaguely recall reason that `ShadowBitmap.drawRect(Rect r, Paint paint)` does not use Graphics2d is because Graphics2d has slight behavior differences. I remember something like if you drawRect with something like 0x00ffffff, which is an 100% traslucent white color, Graphics2d converts it to `0`, and this broke some tests. At least I think that's what it is. Maybe I should capture that in CTesque BitmapTest.

Reply:
For this to work, i  need to create a new method in ShadowBitmap.java with prototype as: 
`void drawRect(int left, int top, int right, int bottom, Paint paint);`
 and then call to this method from shadowCanvas.java

Can i go with it?
what u think @utzcoz

Reply:
I have added a new method named drawRect in ShadowBitmap.java as it was required.
Could you review the changes..

Reply:
@Akshay2131 I have posted new reviewing content.